app/models/inventory_number.rb

```diff
@@ -1,16 +1,29 @@
 class InventoryNumber < ActiveRecord::Base
 
-  def self.all_codes
-    select("CAST(code AS integer) AS number").map(&:number)
+  validates :code, presence: true, uniqueness: true
+
+  def self.create_with_next_code!
+    self.create!(code: next_code)
+  end
+
+  def self.next_code
+    number = missing_code || (max_code + 1)
+    number.to_s.rjust(6, "0")
   end
 
-  def self.latest_code
-    select("CAST(code AS integer)").order("code desc").
-      first.try(:code).try(:to_i) || 1
+  def self.missing_code
+    sql_for_missing_code =
+      "SELECT s.i AS first_missing_code
+        FROM generate_series(1,#{max_code}) s(i)
```

Comment:
Sure @steveyken, Thanks for checking the edge case and I am glad it didn't fail.
i'll update the code and merge it to master.